File: test/scripts/single_thread_cont_test.py
#
# This file is part of libdebug Python library (https://github.com/libdebug/libdebug).
# Copyright (c) 2024 Roberto Alessandro Bertolini, Gabriele Digregorio. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for details.
#

from unittest import TestCase
from utils.binary_utils import RESOLVE_EXE

from libdebug import debugger


def save_thread_state(thread):
    state = b""
    for attr in dir(thread.regs.register_file):
        if not attr.startswith("__"):
            state += getattr(thread.regs.register_file, attr).to_bytes(8, "little")
    return state

class SingleThreadContTest(TestCase):
    def setUp(self):
        self.d = debugger(RESOLVE_EXE("single_thread_cont_test"))
        
    def test_single_thread_continue(self):
        # this function checks that the single-thread cont works correctly. 
        # The other two threads are supposed to be either dead or in a wait.

        self.d.run()

        def callback(_, __):
            pass

        self.d.bp("do_nothing", callback=callback)
        do_nothing_target = self.d.bp("do_nothing_target")

        # This is a process-scoped continue
        self.d.cont()

        thread = None

        for t in self.d.threads:
            if do_nothing_target.hit_on(t):
                # t is our target
                thread = t
                break

        assert thread is not None

        other_threads = self.d.threads.copy()
        other_threads.remove(thread)

        other_threads_state = [save_thread_state(x) for x in other_threads]

        # sanity check
        new_other_threads_state = [save_thread_state(x) for x in other_threads]
        assert all(x == y for x, y in zip(other_threads_state, new_other_threads_state))

        target_state = save_thread_state(thread)

        # calling finish on our target thread should not affect the state of other threads
        thread.cont()
        self.d.wait()

        new_other_threads_state = [save_thread_state(x) for x in other_threads]
        assert all(x == y for x, y in zip(other_threads_state, new_other_threads_state))

        new_target_state = save_thread_state(thread)
        
        assert target_state != new_target_state

        self.d.kill()
        self.d.terminate()

File: test/srcs/single_thread_cont_test.c
//
// This file is part of libdebug Python library (https://github.com/libdebug/libdebug).
// Copyright (c) 2024 Roberto Alessandro Bertolini. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for details.
//

#include <semaphore.h>
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

sem_t sem_other;

void do_nothing()
{

}

void do_nothing_target()
{

}

void do_something_target()
{
    int b = 0;

    for (int i = 0; i < 10; i++) {
        b += i;
    }
}

void* target_thread_function()
{
    do_nothing_target();

    int a = 0;

    for (int i = 0; i < 1024; i++) {
        a += i;
    }

    asm __volatile__("nop");
    do_something_target();
}

void do_nothing_other()
{

}

void* other_thread_function()
{
    int a;
loop:
    a = 0;

    for (int i = 0; i < 1024; i++) {
        do_nothing_other();
        a += i;
    }

    if (sem_trywait(&sem_other)) {
        goto loop;
    }
}

int main()
{
    pthread_t target, other;

    sem_init(&sem_other, 0, 0);

    pthread_create(&target, NULL, target_thread_function, NULL);
    pthread_create(&other, NULL, other_thread_function, NULL);

    do_nothing();

    int a;
    for (int i = 0; i < 10; i++) {
        a += i;
    }

    do_nothing();

    sem_post(&sem_other);

    pthread_join(target, NULL);
    pthread_join(other, NULL);

    return 0;
}